### PR TITLE
Win version for review

### PR DESCRIPTION
**Pull request by rddaz2013**
**caveats** are really the os functions, which does not work the same on linux and windows.

touch()
rename()
makedir()
the operation system file lock functionally
 _(danger! because my portallocker unittest is still not working...ideas for alternatives are welcome)_

**Changes in the py.test...**
rsync .. I have disabled for windows because I had no idea to solve the problem
filesystem.tests - multiprocessing join ..behaves on windows something different. I've reduced the number of test runs in test_death_by_1000_pokes.  My wish would be to have a multiprocessing test without 'join'

The logic and structure of datreant.core is very good, only one thing I had to change. 
Path.endswith (os.sep) in trees.py ...because of the test_leaves/test_tree py.test

Just as for appveyor and Travis CI, I have learned a lot.

### CODE REVIEW

File: appveyor.yml [hunk 1]
```diff
@@ -0,0 +1,57 @@
+environment:
```
Comment (kain88-de) [inline appveyor.yml]: Cool Thanks for that. I'll look into activating it for the PR in the week.

File: .cache/v/cache/lastfailed [hunk 2]
```diff
@@ -0,0 +1,135 @@
+{
```
Comment (kain88-de) [inline .cache/v/cache/lastfailed]: Could you please remove all the hidden dot files you added.

File: src/datreant.core.egg-info/PKG-INFO [hunk 3]
```diff
@@ -0,0 +1,83 @@
+Metadata-Version: 1.1
```
Comment (kain88-de) [inline src/datreant.core.egg-info/PKG-INFO]: Please delete this egg-info as well. It's a build artefact.

File: src/datreant/core/tests/__init__.py [hunk 4]
```diff
@@ -1,15 +0,0 @@
...
-#
-# datreant.tests
-
-"""
-datreant.tests --- unit tests for datreant
-=============================================================
-
-"""
-# Bring some often used objects into the current namespace
-from . import test_collections
```
Comment (kain88-de) [inline src/datreant/core/tests/__init__.py]: Why did you delete this file?

File: src/datreant.core.egg-info/PKG-INFO (hunk 3 above)
Comment (rddaz2013) [inline src/datreant.core.egg-info/PKG-INFO]: best way to remove the trash?

File: src/datreant/core/tests/__init__.py (hunk 4 above)
Comment (rddaz2013) [inline src/datreant/core/tests/__init__.py]: http://python-notes.curiousefficiency.org/en/latest/python_concepts/import_traps.html

Pycharm IDE doesn't liked it while running tests .-)

File: .cache/v/cache/lastfailed (hunk 2 above)
Comment (dotsdl) [inline .cache/v/cache/lastfailed]: Indeed, we don't want these. Not sure of the best way to get rid of them from git history, so easiest way might be to make a new branch off our `develop`, put the files you changed here that want to include directly in there, and then make a new PR against that branch. We can still discuss things here for now though until then.

File: src/datreant/core/trees.py [hunk 5]
```diff
@@ -175,7 +179,7 @@ def __init__(self, dirpath, limbs=None):
             self.attach(*limbs)
 
     def __repr__(self):
-        return "<Tree: '{}'>".format(self.relpath)
+        return "<Tree: '{}'>".format(self.abspath)
```
Comment (dotsdl) [inline src/datreant/core/trees.py]: I'd rather not change the `repr` for `Tree`. Is this necessary?

File: src/datreant/core/trees.py [hunk 6]
```diff
@@ -213,7 +217,7 @@ def __getitem__(self, path):
         def filt(path):
             fullpath = os.path.abspath(os.path.join(self.abspath, path))
 
-            if (os.path.isdir(fullpath) or path.endswith(os.sep) or
+            if (os.path.isdir(fullpath) or path.endswith("/") or
```
Comment (dotsdl) [inline src/datreant/core/trees.py]: `os.sep` is supposed to give the appropriate separator for the given OS. Is this not the case for Windows?

File: src/datreant/core/trees.py [hunk 7]
```diff
@@ -513,8 +516,7 @@ def makedirs(self):
                 this tree
 
         """
-        makedirs(str(self.path))
-
+        makedirs((str(self.path)))
```
Comment (dotsdl) [inline src/datreant/core/trees.py]: ?

File: src/datreant/core/trees.py [hunk 8]
```diff
@@ -130,8 +130,12 @@ def touch(self):
         """Make file if it doesn't exist.
 
         """
-        self.makedirs()
-        self.path.touch()
+        if os.name == 'nt':
```
Comment (dotsdl) [inline src/datreant/core/trees.py]: Looking at the contents of `touch_me`, it's not obvious to me why this is necessary. It looks like it does exactly the same thing?

File: src/datreant/core/util.py [hunk 9]
```diff
@@ -17,3 +20,35 @@ def makedirs(path):
             pass
         else:
             raise
+
+
+def touch_me(path):
```
Comment (dotsdl) [inline src/datreant/core/util.py]: See note above.

File: src/datreant/core/util.py [hunk 10]
```diff
@@ -17,3 +20,35 @@ def makedirs(path):
...
+    Path(path).touch()
+
+
+def fullpath(path):
```
Comment (dotsdl) [inline src/datreant/core/util.py]: What is this function supposed to give? Where is it used?

File: src/datreant/core/util.py [hunk 11]
```diff
@@ -17,3 +20,35 @@ def makedirs(path):
...
+def isfullpath(path):
+    return path == fullpath(path)
+
+
+def path_leaf(path):
+    return os.path.basename(os.path.normpath(path))
+
+
+def relpath(path):
+    """Returns *path* on Windows, and relative path elsewhere."""
```
Comment (dotsdl) [inline src/datreant/core/util.py]: Does a Windows filesystem have no concept of a relative path? Unfortunately I have very little experience with trying to do anything Python on Windows, so apologies for the ignorance on my part.

File: src/datreant/core/util.py [hunk 12]
```diff
@@ -17,3 +20,35 @@ def makedirs(path):
...
+    if not os.path.isabs(path):
+        path = os.path.abspath(path)
+    return path
+
+
+def isfullpath(path):
+    return path == fullpath(path)
+
+
+def path_leaf(path):
```
Comment (dotsdl) [inline src/datreant/core/util.py]: Where is this used?

File: src/datreant/core/util.py [hunk 13]
```diff
@@ -17,3 +20,35 @@ def makedirs(path):
...
+        return None
+    if not isinstance(path, six.string_types):
+        path = str(path)
+    path = os.path.expanduser(path)
+    if not os.path.isabs(path):
+        path = os.path.abspath(path)
+    return path
+
+
+def isfullpath(path):
```
Comment (dotsdl) [inline src/datreant/core/util.py]: Where is this used? If we don't use it, we should remove it.

File: src/datreant/core/limbs.py [hunk 14]
```diff
@@ -3,15 +3,15 @@
...
 
 """
-import os
 import functools
-from six import string_types, with_metaclass
+import os
 from collections import defaultdict
 
 from fuzzywuzzy import process
+from six import string_types, with_metaclass
```
Comment (rddaz2013) [inline src/datreant/core/limbs.py]: pycharm with it's import refactor .-)

File: src/datreant/core/tests/test_locks.py [hunk 15]
```diff
@@ -13,7 +12,7 @@
 def pokefile(treantfilepath, string):
     """Add a number of tags to a Treant."""
     treant = Treant(treantfilepath)
-    treant.tags.add(*["{}_{}".format(string, i) for i in range(100)])
+    treant.tags.add(*["{}_{}".format(string, i) for i in range(10)])
```
Comment (rddaz2013) [inline src/datreant/core/tests/test_locks.py]: Yes a sort of speedup.

join on linux system waits really that all task are ready. In windows the function don't wait...

http://stackoverflow.com/questions/20914828/python-multiprocessing-pool-join-not-waiting-to-go-on

For the 1000 Task my system gets 900 task before the test ask how many results there are..so I reduced the nummer to 100 tasks. Really a WIN think about threads and multitasking, pool.join() is not OS save.

File: src/datreant/core/trees.py (hunk 5 above)
Comment (rddaz2013) [inline src/datreant/core/trees.py]: ? oh...that should be relpath.

relpath on win can't work if two harddisks are used within the path. That why i used a seperate tmpdir for the test.

File: src/datreant/core/trees.py (hunk 6 above)
Comment (rddaz2013) [inline src/datreant/core/trees.py]: The test for tree and treant are posix style, so path.endswith(os.sep) fails on win. 

Design question...should the "dtr[a/path/on/system]..." stay posix-style?

File: src/datreant/core/util.py (hunk 9 above)
Comment (rddaz2013) [inline src/datreant/core/util.py]: PEP8 said 2 lines .-)

File: src/datreant/core/util.py (hunk 11 above)
Comment (rddaz2013) [inline src/datreant/core/util.py]: Np..

https://bugs.python.org/issue7195 
http://stackoverflow.com/questions/4950940/os-path-relpath-not-returning-a-relative-path-when-using-drive-as-start-point

relpath works on windows if not one path is on D:\ and one an E:\ ..between this two there is no relpath. On my Windows system the python env is on D:\ the system on C:\ so i add a tmpdir for the test's 

Also within the windows tmp there a permission issues on WIN10

File: src/datreant.core.egg-info/PKG-INFO (hunk 3 above)
Comment (kain88-de) [inline src/datreant.core.egg-info/PKG-INFO]: https://help.github.com/articles/removing-sensitive-data-from-a-repository/

This helps tells you all you need to know. I used the filter branch script several times myself or to help out colleagues.  You can test it out on a new branch but it should work fine

File: .gitignore [hunk 16]
```diff
@@ -1,3 +1,6 @@
 *.pyc
```
Comment (kain88-de) [inline .gitignore]: @rddaz2013 Can you create  a global gitignore file on windows? I'll use those on linux to keep emacs specific configurations out of the source tree

File: src/datreant/core/tests/test_locks.py (hunk 15 above)
Comment (kain88-de) [inline src/datreant/core/tests/test_locks.py]: Does this code spin up other processes?

File: appveyor.yml (hunk 1 above)
Comment (kain88-de) [inline appveyor.yml]: Appveyor should be actiated now. Could be that we need to reopen this PR for the builds to be started or maybe a push is enough.

File: src/datreant/core/trees.py (hunk 6 above)
Comment (dotsdl) [inline src/datreant/core/trees.py]: Oh I see, yeah if the posix-style paths work fine on Windows, let's keep this, then.

File: src/datreant/core/util.py (hunk 9 above)
Comment (dotsdl) [inline src/datreant/core/util.py]: I was more asking about what `touch_me` does. It looks to me like it does the same thing as the code it is intended to replace. What's the point?

File: setup.cfg [hunk 17]
```diff
@@ -1,5 +1,6 @@
 [pytest]
 pep8ignore = __init__.py E402
+-addopts = -v --basetemp=E:\mydir
```
Comment (rddaz2013) [inline setup.cfg]: Yes it's only for windows..

On one of my dev-environment the python interpreter runs on E:\ but the tmp on win10 is on C:\ that runs not good.

The reasons therefore I could not find.
a) is the problem the relpath thing on win
b) User/Admin Permissions on win10 (env is on a workplace where i am not admin-rights!)

 Can pytest differentiate between win/linux for the cfg?

**Comment by kain88-de** [comment]
@rddaz2013 I think we will take care of this in the spring cleaning. That changes a whole lot of code anyway. Thanks for getting this started.
